src/Database/LSMTree/Internal/Index.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE TypeFamilies #-}
+
+{-|
+    Provides a common interface to different types of fence pointer indexes and
+    their accumulators.
+-}
+module Database.LSMTree.Internal.Index
+(
+    Index (search, sizeInPages, fromSBS),
+    IndexAcc (ResultingIndex, appendSingle, appendMulti, unsafeEnd)
+)
+where
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.ByteString.Short (ShortByteString)
+import           Data.Word (Word32)
+import           Database.LSMTree.Internal.Chunk (Chunk)
+import           Database.LSMTree.Internal.Entry (NumEntries)
+import           Database.LSMTree.Internal.Page (NumPages, PageSpan)
+import           Database.LSMTree.Internal.Serialise (SerialisedKey)
+
+-- | The class of index types.
+class Index i where
+
+    {-|
+        Searches for a page span that contains a key–value pair with the given
+        key. If there is indeed such a pair, the result is the corresponding
+        page span; if there is no such pair, the result is an arbitrary but
+        valid page span.
+    -}
+    search :: SerialisedKey -> i -> PageSpan
+
+    -- | Yields the number of pages covered by an index.
+    sizeInPages :: i -> NumPages
+
+    {-|
+        Reads an index along with the number of entries of the respective run
+        from a byte string.
+
+        The byte string must contain the serialised index exactly, with no
+        leading or trailing space.
+    -}
+    fromSBS :: ShortByteString -> Either String (NumEntries, i)
+
+{-|
+    The class of index accumulator types, where an index accumulator denotes an
+    index under incremental construction.
+
+    Incremental index construction is only guaranteed to work correctly when the
+    following conditions are met:
+
+      * The provided information about pages is not in conflict with the
+        requirement that the keys in a run are ascending.
+
+      * The length of each key can be represented by a 16-bit word.
+-}
+class Index (ResultingIndex j) => IndexAcc j where
```

Comment:
Shouldn't `new` also be part of the class?

Reply:
Originally, I had intended to have `new` as a class method, but then I dropped this idea.

An issue with `new` is that its different implementations do not take the same kind of data via arguments. Yes, the versions for the compact and the ordinary index both have an `Int` argument. However, this argument means different things depending on the index type: for compact indexes it specifies the *maximum* chunk size, for ordinary indexes it specifies the *minimum* chunk size. It is conceivable that in the future another index type is added whose implementation of `new` takes data of an entirely different kind.

Of course, it would be possible to use an associated type to specify what the type of the input to `new` should be. However, this would be unnecessarily complicated, as a polymorphic `new` isn’t really needed. I envision that the type of index and perhaps concrete data to feed to `new` is received as a configuration and then a specific implementation of `new` is invoked depending on this configuration. After an index accumulator has been produced this way, constructing and querying the index will be performed using the class methods.

Reply:
Note that there seems to be an analogy to object-oriented programming in C++ here: in C++, ordinary methods and destructors can be virtual, allowing dynamic binding, but constructors can’t.

Reply:
All right, I suppose we'll see in a later PR how `new` is used